src/peripheral/mod.rs

```diff
@@ -100,13 +104,59 @@ pub struct Cpuid {
     pub isar: [RO<u32>; 5],
     reserved1: u32,
     /// Cache Level ID
+    #[cfg(armv7m)]
     pub clidr: RO<u32>,
     /// Cache Type
+    #[cfg(armv7m)]
     pub ctr: RO<u32>,
     /// Cache Size ID
+    #[cfg(armv7m)]
     pub ccsidr: RO<u32>,
     /// Cache Size Selection
-    pub csselr: RO<u32>,
+    #[cfg(armv7m)]
+    pub csselr: RW<u32>,
+}
+
+const CSSELR_IND_POS: u32 = 0;
+const CSSELR_IND_MASK: u32 = 1 << CSSELR_IND_POS;
+const CSSELR_LEVEL_POS: u32 = 1;
+const CSSELR_LEVEL_MASK: u32 = 0x7 << CSSELR_LEVEL_POS;
+const CCSIDR_NUMSETS_POS: u32 = 13;
+const CCSIDR_NUMSETS_MASK: u32 = 0x7FFF << CCSIDR_NUMSETS_POS;
+const CCSIDR_ASSOCIATIVITY_POS: u32 = 3;
+const CCSIDR_ASSOCIATIVITY_MASK: u32 = 0x3FF << CCSIDR_ASSOCIATIVITY_POS;
+
+/// Type of cache to select on CSSELR writes.
+#[cfg(armv7m)]
+pub enum CsselrCacheType {
+    /// Select DCache or unified cache
+    DataOrUnified = 0,
+    /// Select ICache
+    Instruction = 1,
+}
+
+#[cfg(armv7m)]
+impl Cpuid {
+    /// Selects the current CCSIDR
+    ///
+    /// * `level`: the required cache level minus 1, e.g. 0 for L1, 1 for L2
+    /// * `ind`: select instruction cache or data/unified cache
+    pub fn select_cache(&self, level: u8, ind: CsselrCacheType) {
+        assert!(level<8);
+        unsafe { self.csselr.write(
+            (((level as u32) << CSSELR_LEVEL_POS) & CSSELR_LEVEL_MASK) |
```

Comment:
Hmm, sure. It can't happen here because level is a u8 but it could happen for `way` in the CBP methods. I would rather keep the `_MASK` constants as shifted, but we can just shift them down again before masking and it'l compile to the same deal.